### ISSUE TITLE
how to prepare customized training data for retriever

### ISSUE THREAD

**Issue by luomancs** [facebookresearch/DPR#91]
Hi, 

I want to use my dataset to train a retriever to fit my domain problem. So I looked into the nq-dev.json and found that in the positive_ctxs field, for each passage, there is a  score and a title score,  how the score is computed, and will they be used in the training time? Also, I found negative_ctxs and hard_negative_ctxs, what are the differences between them, and how the score and title_score are computed? 

Lastly, should I have the same number of positive_ctxs for all questions? and how about the number of hard_negative_ctxs and negative_ctxs. 

thank you

**Comment by vlad-karpukhin** [comment]
Hi there, 
sorry for a long reply
The score of the passages are the bm25 scores  but we don't use them at the moment. You can just skip them.
negative_ctxs -  some random negative passages form entire corpora (we don't use them by default, see https://github.com/facebookresearch/DPR/blob/master/train_dense_encoder.py#L520)
hard_negative_ctxs - negative passages from the top 30-50 BM25 results.
The current code uses just 1 passages in the positive_ctxs. This selection is randomized for train samples if --shuffle_positives cli param is activated

**Comment by unw9527** [comment]
Hi @vlad-karpukhin , if I want to train on my own dataset, how should I modify the command line here?
`train_datasets=[list of train datasets, comma separated without spaces]`
Thanks a lot!

**Comment by vlad-karpukhin** [comment]
Hi @unw9527 ,
Please read the configuration logic (https://github.com/facebookresearch/DPR/tree/main/conf).
What you need to do is to define your custom data loader like dpr.data.biencoder_data.JsonQADataset and put the configuration at conf/datasets/encoder_train_default.yaml or your own configuration file.
You can put just absolute file path in your FS for the 'file' attribute.

**Comment by vlad-karpukhin** [comment]
The keyword you will use to configure your dataset (see 'nq_train' in conf/datasets/encoder_train_default.yaml, for example) will be the key to use in the train_datasets lists, but please always use the list notation for train_datasets, even in the case of a single dataset.
